src/bcif.jl

```diff
@@ -0,0 +1,643 @@
+    end
+    params = content
+
+    encoding_constructors = Dict(
+        "ByteArray" => () -> ByteArrayEncoding(INT_TO_TYPE[get(params, "type", nothing)]),
+        "FixedPoint" => () -> FixedPointEncoding(params["factor"]; srcType=INT_TO_TYPE[get(params, "srcType", FLOAT32)]),
+        "IntervalQuantization" => () -> IntervalQuantizationEncoding(params["min"], params["max"], params["numSteps"];
+            srcType=INT_TO_TYPE[get(params, "srcType", 32)]),
+        "RunLength" => () -> RunLengthEncoding(srcSize=get(params, "srcSize", nothing),
+            srcType=INT_TO_TYPE[get(params, "srcType", nothing)]),
+        "Delta" => () -> DeltaEncoding(srcType=INT_TO_TYPE[get(params, "srcType", nothing)],
+            origin=get(params, "origin", 0)),
+        "IntegerPacking" => () -> IntegerPackingEncoding(params["byteCount"],
+            srcSize=get(params, "srcSize", nothing),
+            isUnsigned=get(params, "isUnsigned", false))
+    )
+
+    if haskey(encoding_constructors, kind)
+        return encoding_constructors[kind]()
+    else
+        error("Unknown encoding kind: $kind")
+    end
+end
+
+
+function decode_column(column::Dict)
+    column_data = column["data"]
+    encodings = []
+
+    # collect the encodings. If it's a string encoding then it should be a single encoding
+    # that contains it's own dataEncoding and offsetEncoding which also need to be handled
+    for enc in column_data["encoding"]
+        if enc["kind"] == "StringArray"
+            push!(encodings, StringArrayEncoding(
+                stringData=enc["stringData"],
+                dataEncoding=deserialize_numeric_encoding(enc["dataEncoding"]),
+                offsetEncoding=deserialize_numeric_encoding(enc["offsetEncoding"]),
+                offsets=enc["offsets"]
+            ))
+        else
+            push!(encodings, deserialize_numeric_encoding(enc))
+        end
+    end
+
+    return decode_stepwise(column_data["data"], encodings)
+end
+
+
+# Below are the encoding and decoding types for BCIF format
+
+# Data types defined for the BCIF encoding by are indicated by integer values
+# there are not well discussed in the official spec, had to ask about it excplicitly
+# https://github.com/molstar/BinaryCIF/issues/4
+@enum TypeCode begin
+    INT8 = 1
+    INT16 = 2
+    INT32 = 3
+    UINT8 = 4
+    UINT16 = 5
+    UINT32 = 6
+    FLOAT32 = 32
+    FLOAT64 = 33
+end
+
+const INT_TO_TYPE = Dict(
+    1 => Int8,
+    2 => Int16,
+    3 => Int32,
+    4 => UInt8,
+    5 => UInt16,
+    6 => UInt32,
+    32 => Float32,
+    33 => Float64
+)
+
+# Mapping from TypeCode to Julia types
+const TYPE_CODE_TO_TYPE = Dict(
+    INT8 => Int8,
+    INT16 => Int16,
+    INT32 => Int32,
+    UINT8 => UInt8,
+    UINT16 => UInt16,
+    UINT32 => UInt32,
+    FLOAT32 => Float32,
+    FLOAT64 => Float64
+)
+
+# Mapping from Julia types to TypeCode
+const TYPE_TO_TYPE_CODE = Dict(value => key for (key, value) in TYPE_CODE_TO_TYPE)
```

Comment:
```suggestion
const TYPE_TO_TYPE_CODE = IdDict{Any,TypeCode}(value => key for (key, value) in TYPE_CODE_TO_TYPE)
```
When using types as keys, `IdDict` is much faster than `Dict`.
Another advantage is that you won't have to `rehash!` it, see https://docs.julialang.org/en/v1/manual/modules/#Module-initialization-and-precompilation (and then search for `IdDict`).